### PR TITLE
Implemented integration test focused on the migration between registries of upkeeps

### PR DESCRIPTION
**Pull request by vla33**

### CODE REVIEW

File: integration-tests/smoke/keeper_test.go [hunk 1]
```diff
@@ -585,6 +587,79 @@ func getKeeperSuite(
...
+
+					// Set the jobs for the second registry
+					actions.CreateKeeperJobs(chainlinkNodes, secondRegistry)
+					err = c.WaitForEvents()
+					Expect(err).ShouldNot(HaveOccurred(), "Error creating keeper jobs")
+
+					err := registry.SetMigrationPermissions(common.HexToAddress(registry.Address()), 3)
+					Expect(err).ShouldNot(HaveOccurred(), "Couldn't set bidirectional permissions for first registry")
+					err = c.WaitForEvents()
+					Expect(err).ShouldNot(HaveOccurred(), "Failed to wait for migration permissions setting")
```
Comment (infiloop2) [inline integration-tests/smoke/keeper_test.go:613]: nit. you can remove this and wait for both migration txs at once

File: integration-tests/smoke/keeper_test.go [hunk 2]
```diff
@@ -585,6 +587,79 @@ func getKeeperSuite(
...
+						c,
+						linkFundsForEachUpkeep,
+					)
+
+					// Set the jobs for the second registry
+					actions.CreateKeeperJobs(chainlinkNodes, secondRegistry)
+					err = c.WaitForEvents()
+					Expect(err).ShouldNot(HaveOccurred(), "Error creating keeper jobs")
+
+					err := registry.SetMigrationPermissions(common.HexToAddress(registry.Address()), 3)
```
Comment (infiloop2) [inline integration-tests/smoke/keeper_test.go:610]:
```suggestion
					err := registry.SetMigrationPermissions(common.HexToAddress(secondRegistry.Address()), 3)
```

File: integration-tests/smoke/keeper_test.go [hunk 3]
```diff
@@ -585,6 +587,79 @@ func getKeeperSuite(
...
+					Expect(err).ShouldNot(HaveOccurred(), "Couldn't set bidirectional permissions for first registry")
+					err = c.WaitForEvents()
+					Expect(err).ShouldNot(HaveOccurred(), "Failed to wait for migration permissions setting")
+
+					err = registry.SetMigrationPermissions(common.HexToAddress(secondRegistry.Address()), 3)
```
Comment (infiloop2) [inline integration-tests/smoke/keeper_test.go:615]:
```suggestion
					err = secondRegistry.SetMigrationPermissions(common.HexToAddress(registry.Address()), 3)
```

File: integration-tests/smoke/cron_test.go [hunk 4]
```diff
@@ -83,7 +83,7 @@ var _ = Describe("Cronjob suite @cron", func() {
 
 	AfterEach(func() {
 		By("Tearing down the environment", func() {
-			err = actions.TeardownSuite(e, utils.ProjectRoot, []client.Chainlink{chainlinkNode}, nil)
+			err = actions.TeardownSuite(e, utils.ProjectRoot, []client.Chainlink{chainlinkNode}, nil, nil)
```
Comment (infiloop2) [inline integration-tests/smoke/cron_test.go:86]: nit. unrelated

File: integration-tests/smoke/keeper_test.go [hunk 5]
```diff
@@ -585,6 +587,79 @@ func getKeeperSuite(
...
+						counterBeforeMigration, err = consumers[0].Counter(context.Background())
+						g.Expect(err).ShouldNot(HaveOccurred(), "Calling consumer's counter shouldn't fail")
+						g.Expect(counterBeforeMigration.Int64()).Should(BeNumerically(">", int64(0)),
+							"Expected consumer counter to be greater than 0, but got %s", counterBeforeMigration)
+					}, "1m", "1s").Should(Succeed())
+
+					upkeepInfo, err := registry.GetUpkeepInfo(context.Background(), upkeepIDs[0])
+					Expect(err).ShouldNot(HaveOccurred(), "Couldn't retrieve upkeep info")
+					err = c.WaitForEvents()
+					Expect(err).ShouldNot(HaveOccurred(), "Failed to wait for upkeep info retrieval")
```
Comment (infiloop2) [inline integration-tests/smoke/keeper_test.go:633]: you don't need to wait for reads

File: integration-tests/smoke/keeper_test.go [hunk 6]
```diff
@@ -585,6 +587,79 @@ func getKeeperSuite(
...
+
+					// Grant permission to the registry to fund the upkeep
+					err = linkToken.Approve(registry.Address(), upkeepInfo.Balance)
+					Expect(err).ShouldNot(HaveOccurred(), "Failed to approve")
+					err = c.WaitForEvents()
+					Expect(err).ShouldNot(HaveOccurred(), "Failed to wait for events")
+
+					// Migrate the upkeep with ID 0 from the first to the second registry
```
Comment (infiloop2) [inline integration-tests/smoke/keeper_test.go:641]: nit.
```suggestion
					// Migrate the upkeep with index 0 from the first to the second registry
```

File: integration-tests/smoke/keeper_test.go [hunk 7]
```diff
@@ -585,6 +587,79 @@ func getKeeperSuite(
...
+					upkeepInfo, err := registry.GetUpkeepInfo(context.Background(), upkeepIDs[0])
+					Expect(err).ShouldNot(HaveOccurred(), "Couldn't retrieve upkeep info")
+					err = c.WaitForEvents()
+					Expect(err).ShouldNot(HaveOccurred(), "Failed to wait for upkeep info retrieval")
+
+					// Grant permission to the registry to fund the upkeep
+					err = linkToken.Approve(registry.Address(), upkeepInfo.Balance)
+					Expect(err).ShouldNot(HaveOccurred(), "Failed to approve")
+					err = c.WaitForEvents()
+					Expect(err).ShouldNot(HaveOccurred(), "Failed to wait for events")
```
Comment (infiloop2) [inline integration-tests/smoke/keeper_test.go:639]: nit. shouldn't be needed

File: integration-tests/smoke/keeper_test.go [hunk 8]
```diff
@@ -585,6 +587,79 @@ func getKeeperSuite(
...
+					Expect(err).ShouldNot(HaveOccurred(), "Couldn't migrate the first upkeep")
+					err = c.WaitForEvents()
+					Expect(err).ShouldNot(HaveOccurred(), "Failed to wait for migration")
+
+					// Pause the first registry, in that way we make sure that the upkeep is being performed by the second one
+					err = registry.Pause()
+					Expect(err).ShouldNot(HaveOccurred(), "Could not pause the registry")
+					err = c.WaitForEvents()
+					Expect(err).ShouldNot(HaveOccurred(), "Failed to wait for the pausing of the first registry")
+
```
Comment (infiloop2) [inline integration-tests/smoke/keeper_test.go:652]: can we refetch `counterAfterMigration` here and then check in eventually block that it increases?. It might have been performed more times in between all these txs

File: integration-tests/smoke/keeper_test.go [hunk 9]
```diff
@@ -585,6 +587,79 @@ func getKeeperSuite(
...
+					err = c.WaitForEvents()
+					Expect(err).ShouldNot(HaveOccurred(), "Failed to wait for events")
+
+					// Migrate the upkeep with ID 0 from the first to the second registry
+					err = registry.Migrate([]*big.Int{upkeepIDs[0]}, common.HexToAddress(secondRegistry.Address()))
+					Expect(err).ShouldNot(HaveOccurred(), "Couldn't migrate the first upkeep")
+					err = c.WaitForEvents()
+					Expect(err).ShouldNot(HaveOccurred(), "Failed to wait for migration")
+
+					// Pause the first registry, in that way we make sure that the upkeep is being performed by the second one
```
[the remark this answers is not in the document]
Comment (vla33) [reply integration-tests/smoke/keeper_test.go:635]: Thanks Akshay! 😊

File: integration-tests/smoke/keeper_test.go (hunk 8 above)
Comment (vla33) [reply integration-tests/smoke/keeper_test.go:640]: Sure, updated it!

File: integration-tests/go.mod [hunk 10]
```diff
@@ -8,8 +8,8 @@ require (
 	github.com/onsi/gomega v1.19.0
 	github.com/rs/zerolog v1.27.0
 	github.com/satori/go.uuid v1.2.0
-	github.com/smartcontractkit/chainlink-env v0.2.18
-	github.com/smartcontractkit/chainlink-testing-framework v1.4.2
+	github.com/smartcontractkit/chainlink-env v0.2.19
+	github.com/smartcontractkit/chainlink-testing-framework v1.4.3-0.20220701101352-08f4f955cdf4
```
Comment (jmank88) [inline integration-tests/go.mod:12]: Is it possible to stick to a tagged semver release instead?

Comment (vla33) [reply integration-tests/go.mod:12]: The changes made in this PR are dependent on some new additions in the `chainlink-testing-framework` repo (the ones included in the has of the commit indicated on line 12), so we'd need a new release from the QA team, which I don't know if they would do just for these changes. Is it bad practice to point to a commit like I do at the moment?

**Review by infiloop2 — APPROVED** [review]
lgtm! Thanks for adding this

File: integration-tests/go.mod (hunk 10 above)
Comment (jmank88) [reply integration-tests/go.mod:12]: It is not necessarily problematic, but it is harder for us to track as developers and may overly constrain how the tooling resolves the versions. But regardless, It looks like github.com/smartcontractkit/chainlink-testing-framework is on 1.4.6 now, no? Shouldn't this either be included in a tagged version, or a variant of the latest version?

Comment (infiloop2) [reply integration-tests/go.mod:12]: I think this is expected since we are doing these changes often in testing framework. 
Not sure @kalverra if you have any better solution in mind?

Comment (vla33) [reply integration-tests/go.mod:12]: Apologies Jordan, I wasn't sure when you're going to reply and I merged the PR before I saw your reply...I understand, that makes sense, thank you for letting me know! I am going to do a refactoring anyway today for the Keeper tests, so I can change the version then to a proper one. Would that be all right? In a couple of hours or so.

Comment (jmank88) [reply integration-tests/go.mod:12]: No bother - sergey actually already has another bump in flight here: #6909

Comment (vla33) [reply integration-tests/go.mod:12]: Oh great, that's even better then! Thanks!
